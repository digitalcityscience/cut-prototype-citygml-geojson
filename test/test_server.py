from flask.testing import FlaskClient
import pathmagic
import urllib.parse
assert pathmagic


def test_endpoint_no_parameters_retuns_all(client: FlaskClient, snapshot):
    response = client.get('/features')

    assert response.status_code == 200
    assert response.json == snapshot


def test_endpoint_with_bbox_returns_empty(client: FlaskClient, snapshot) -> None:
    response = client.get(
        '/features?bbox=10.039595125418998,53.58188067839408,10.042624874581003,53.58367932160592')

    assert response.status_code == 200
    assert response.json == snapshot


def test_endpoint_with_bbox_returns_correct_features(client: FlaskClient, snapshot) -> None:
    response = client.get(
        '/features?bbox=10.315909,53.446486,10.316386,53.446842')

    assert response.status_code == 200
    actual = response.json
    actual["features"] = [f["properties"]["id"] for f in actual["features"]]
    assert actual == snapshot


def test_endpoint_with_bbox_enclosing_all_returns_all(client: FlaskClient, snapshot) -> None:
    response = client.get('/features?bbox=-180,-90,180,90')
    assert response.status_code == 200
    actual = response.json
    actual["features"] = [f["properties"]["id"] for f in actual["features"]]
    assert actual == snapshot


def test_endpoint_with_limit(client: FlaskClient, snapshot) -> None:
    response = client.get('/features?bbox=-180,-90,180,90&limit=1&startindex=0')

    assert response.status_code == 200
    actual = response.json
    actual["features"] = [f["properties"]["id"] for f in actual["features"]]
    assert actual == snapshot


def test_endpoint_with_limit_returns_prev_and_next(client: FlaskClient, snapshot) -> None:
    response = client.get('/features?bbox=-180,-90,180,90&limit=1&startindex=1')

    assert response.status_code == 200
    actual = response.json
    actual["features"] = [f["properties"]["id"] for f in actual["features"]]
    assert actual == snapshot


def test_endpoint_with_filter_ground(client: FlaskClient) -> None:
    response = client.get('/features?surface_type=ground')

    assert response.status_code == 200
    actual = response.json
    assert (len(actual["features"])) == 2


def test_endpoint_with_filter_roof(client: FlaskClient) -> None:
    response = client.get('/features?surface_type=roof')

    assert response.status_code == 200
    actual = response.json
    assert (len(actual["features"])) == 3


def test_endpoint_with_filter_id(client: FlaskClient) -> None:
    response = client.get("/features?properties=id")

    assert response.status_code == 200
    actual = response.json
    assert actual["features"][0]["properties"] == {
        "id": "DEHH_2cfcae78-56f5-4a77-b74d-51cce7cc3d3e_2_poly"
    }


def test_endpoint_with_filter_id_and_lage(client: FlaskClient) -> None:
    response = client.get("/features?properties=id,lage")

    assert response.status_code == 200
    actual = response.json
    assert actual["features"][0]["properties"] == {
        "id": "DEHH_2cfcae78-56f5-4a77-b74d-51cce7cc3d3e_2_poly",
        "lage": "1000",
    }


def test_endpoint_with_filter_invalid_property(client: FlaskClient) -> None:
    response = client.get("/features?properties=id,invalid")

    assert response.status_code == 200
    actual = response.json
    assert actual["features"][0]["properties"] == {
        "id": "DEHH_2cfcae78-56f5-4a77-b74d-51cce7cc3d3e_2_poly",
    }


def test_endpoint_with_bpoly_returns_correct_features(client: FlaskClient) -> None:
    polygon_str = "10.315909,53.446486;10.315909,53.446842;10.316386,53.446842;10.316386,53.446486;10.315909,53.446486"
    response = client.get(f'/features?bounding_polygon={polygon_str}')

    assert response.status_code == 200
    actual = response.json
    ids = [f["properties"]["id"] for f in actual["features"]]
    assert ids == [
        'DEHH_2cfcae78-56f5-4a77-b74d-51cce7cc3d3e_2_poly',
        'DEHH_d1c15e0c-707a-4734-abea-2a1f810c1890_2_poly',
        'DEHH_1fd3c3ef-fc46-4024-ac10-3b272044391d_2_poly']


def test_endpoint_with_bpoly_urlencoded(client: FlaskClient) -> None:
    polygon_str = "10.315909,53.446486;10.315909,53.446842;10.316386,53.446842;10.316386,53.446486;10.315909,53.446486"
    encoded_polygon = urllib.parse.quote(polygon_str)
    url = f"/features?bounding_polygon={encoded_polygon}"

    response = client.get(url)

    assert response.status_code == 200
    actual = response.json
    ids = [f["properties"]["id"] for f in actual["features"]]
    assert ids == [
        'DEHH_2cfcae78-56f5-4a77-b74d-51cce7cc3d3e_2_poly',
        'DEHH_d1c15e0c-707a-4734-abea-2a1f810c1890_2_poly',
        'DEHH_1fd3c3ef-fc46-4024-ac10-3b272044391d_2_poly']


def test_endpoint_with_intersect(client: FlaskClient, snapshot) -> None:
    response = client.get(
        '/features?bbox=10.315909,53.446486,10.316386,53.446842&intersect=true')

    assert response.status_code == 200
    actual = response.json
    assert actual == snapshot


def test_endpoint_with_bpoly_and_intersect(client: FlaskClient, snapshot) -> None:
    polygon_str = "10.315909,53.446486;10.315909,53.446842;10.316386,53.446842;10.316386,53.446486;10.315909,53.446486"
    response = client.get(f'/features?bounding_polygon={polygon_str}&intersect=true')

    assert response.status_code == 200
    actual = response.json
    assert actual == snapshot
