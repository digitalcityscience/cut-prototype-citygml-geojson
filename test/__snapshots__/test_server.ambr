# serializer version: 1
# name: test_endpoint_no_parameters_retuns_all
  dict({
    'features': list([
      dict({
        'geometry': dict({
          'coordinates': list([
            list([
              list([
                10.316243,
                53.446842,
              ]),
              list([
                10.316386,
                53.446785,
              ]),
              list([
                10.316051,
                53.446486,
              ]),
              list([
                10.315909,
                53.446543,
              ]),
              list([
                10.316243,
                53.446842,
              ]),
            ]),
          ]),
          'type': 'Polygon',
        }),
        'properties': dict({
          'Geometrietyp2DReferenz': '3000',
          'Grundrissaktualitaet': '2023-01-26',
          'bodenhoehe': '1100',
          'dachhoehe': '1000',
          'function': '31001_3070',
          'geschossanzahl': '1000',
          'id': 'DEHH_2cfcae78-56f5-4a77-b74d-51cce7cc3d3e_2_poly',
          'lage': '1000',
          'measuredHeight': '4.627',
          'roofType': '3100',
          'storeysAboveGround': '1',
          'surface_type': 'ground',
        }),
        'type': 'Feature',
      }),
      dict({
        'geometry': dict({
          'coordinates': list([
            list([
              list([
                10.317512,
                53.448376,
              ]),
              list([
                10.317382,
                53.448218,
              ]),
              list([
                10.317378,
                53.44822,
              ]),
              list([
                10.317307,
                53.448241,
              ]),
              list([
                10.317437,
                53.448399,
              ]),
              list([
                10.317508,
                53.448378,
              ]),
              list([
                10.317512,
                53.448376,
              ]),
            ]),
          ]),
          'type': 'Polygon',
        }),
        'properties': dict({
          'Geometrietyp2DReferenz': '3000',
          'Grundrissaktualitaet': '2023-01-26',
          'bodenhoehe': '1100',
          'dachhoehe': '1000',
          'function': '31001_2140',
          'geschossanzahl': '1000',
          'id': 'DEHH_5f54fbe2-c26d-4454-bf10-78b218e6d451_2_poly',
          'lage': '1000',
          'measuredHeight': '9.000',
          'roofType': '1000',
          'storeysAboveGround': '1',
          'surface_type': 'ground',
        }),
        'type': 'Feature',
      }),
      dict({
        'geometry': dict({
          'coordinates': list([
            list([
              list([
                10.31598,
                53.446514,
              ]),
              list([
                10.316314,
                53.446814,
              ]),
              list([
                10.316243,
                53.446842,
              ]),
              list([
                10.315909,
                53.446543,
              ]),
              list([
                10.31598,
                53.446514,
              ]),
            ]),
          ]),
          'type': 'Polygon',
        }),
        'properties': dict({
          'id': 'DEHH_d1c15e0c-707a-4734-abea-2a1f810c1890_2_poly',
          'surface_type': 'roof',
        }),
        'type': 'Feature',
      }),
      dict({
        'geometry': dict({
          'coordinates': list([
            list([
              list([
                10.316051,
                53.446486,
              ]),
              list([
                10.316386,
                53.446785,
              ]),
              list([
                10.316314,
                53.446814,
              ]),
              list([
                10.31598,
                53.446514,
              ]),
              list([
                10.316051,
                53.446486,
              ]),
            ]),
          ]),
          'type': 'Polygon',
        }),
        'properties': dict({
          'id': 'DEHH_1fd3c3ef-fc46-4024-ac10-3b272044391d_2_poly',
          'surface_type': 'roof',
        }),
        'type': 'Feature',
      }),
      dict({
        'geometry': dict({
          'coordinates': list([
            list([
              list([
                10.317508,
                53.448378,
              ]),
              list([
                10.317437,
                53.448399,
              ]),
              list([
                10.317307,
                53.448241,
              ]),
              list([
                10.317378,
                53.44822,
              ]),
              list([
                10.317382,
                53.448218,
              ]),
              list([
                10.317512,
                53.448376,
              ]),
              list([
                10.317508,
                53.448378,
              ]),
            ]),
          ]),
          'type': 'Polygon',
        }),
        'properties': dict({
          'id': 'DEHH_4168b264-76af-4aa2-abf7-fafd4b168b7a_2_poly',
          'surface_type': 'roof',
        }),
        'type': 'Feature',
      }),
    ]),
    'links': list([
    ]),
    'numberMatched': 5,
    'numberReturned': 5,
    'timeStamp': '2024-01-01T12:00:01Z',
    'type': 'FeatureCollection',
  })
# ---
# name: test_endpoint_with_bbox_enclosing_all_returns_all
  dict({
    'features': list([
      'DEHH_2cfcae78-56f5-4a77-b74d-51cce7cc3d3e_2_poly',
      'DEHH_5f54fbe2-c26d-4454-bf10-78b218e6d451_2_poly',
      'DEHH_d1c15e0c-707a-4734-abea-2a1f810c1890_2_poly',
      'DEHH_1fd3c3ef-fc46-4024-ac10-3b272044391d_2_poly',
      'DEHH_4168b264-76af-4aa2-abf7-fafd4b168b7a_2_poly',
    ]),
    'links': list([
    ]),
    'numberMatched': 5,
    'numberReturned': 5,
    'timeStamp': '2024-01-01T12:00:01Z',
    'type': 'FeatureCollection',
  })
# ---
# name: test_endpoint_with_bbox_returns_correct_features
  dict({
    'features': list([
      'DEHH_2cfcae78-56f5-4a77-b74d-51cce7cc3d3e_2_poly',
      'DEHH_d1c15e0c-707a-4734-abea-2a1f810c1890_2_poly',
      'DEHH_1fd3c3ef-fc46-4024-ac10-3b272044391d_2_poly',
    ]),
    'links': list([
    ]),
    'numberMatched': 3,
    'numberReturned': 3,
    'timeStamp': '2024-01-01T12:00:01Z',
    'type': 'FeatureCollection',
  })
# ---
# name: test_endpoint_with_bbox_returns_empty
  dict({
    'features': list([
    ]),
    'links': list([
    ]),
    'numberMatched': 0,
    'numberReturned': 0,
    'timeStamp': '2024-01-01T12:00:01Z',
    'type': 'FeatureCollection',
  })
# ---
# name: test_endpoint_with_bpoly_and_intersect
  dict({
    'features': list([
      dict({
        'geometry': dict({
          'coordinates': list([
            list([
              list([
                10.316386,
                53.44678499999999,
              ]),
              list([
                10.316051,
                53.446486,
              ]),
              list([
                10.315909,
                53.44654299999999,
              ]),
              list([
                10.316243,
                53.44684199999999,
              ]),
              list([
                10.316386,
                53.44678499999999,
              ]),
            ]),
          ]),
          'type': 'Polygon',
        }),
        'properties': dict({
          'Geometrietyp2DReferenz': '3000',
          'Grundrissaktualitaet': '2023-01-26',
          'bodenhoehe': '1100',
          'dachhoehe': '1000',
          'function': '31001_3070',
          'geschossanzahl': '1000',
          'id': 'DEHH_2cfcae78-56f5-4a77-b74d-51cce7cc3d3e_2_poly',
          'lage': '1000',
          'measuredHeight': '4.627',
          'roofType': '3100',
          'storeysAboveGround': '1',
          'surface_type': 'ground',
        }),
        'type': 'Feature',
      }),
      dict({
        'geometry': dict({
          'coordinates': list([
            list([
              list([
                10.316314,
                53.446814,
              ]),
              list([
                10.31598,
                53.446514,
              ]),
              list([
                10.315909,
                53.44654299999999,
              ]),
              list([
                10.316243,
                53.44684199999999,
              ]),
              list([
                10.316314,
                53.446814,
              ]),
            ]),
          ]),
          'type': 'Polygon',
        }),
        'properties': dict({
          'id': 'DEHH_d1c15e0c-707a-4734-abea-2a1f810c1890_2_poly',
          'surface_type': 'roof',
        }),
        'type': 'Feature',
      }),
      dict({
        'geometry': dict({
          'coordinates': list([
            list([
              list([
                10.316051,
                53.446486,
              ]),
              list([
                10.31598,
                53.446514,
              ]),
              list([
                10.316314,
                53.446814,
              ]),
              list([
                10.316386,
                53.44678499999999,
              ]),
              list([
                10.316051,
                53.446486,
              ]),
            ]),
          ]),
          'type': 'Polygon',
        }),
        'properties': dict({
          'id': 'DEHH_1fd3c3ef-fc46-4024-ac10-3b272044391d_2_poly',
          'surface_type': 'roof',
        }),
        'type': 'Feature',
      }),
    ]),
    'links': list([
    ]),
    'numberMatched': 3,
    'numberReturned': 3,
    'timeStamp': '2024-01-01T12:00:01Z',
    'type': 'FeatureCollection',
  })
# ---
# name: test_endpoint_with_intersect
  dict({
    'features': list([
      dict({
        'geometry': dict({
          'coordinates': list([
            list([
              list([
                10.316386,
                53.44678499999999,
              ]),
              list([
                10.316051,
                53.446486,
              ]),
              list([
                10.315909,
                53.44654299999999,
              ]),
              list([
                10.316243,
                53.44684199999999,
              ]),
              list([
                10.316386,
                53.44678499999999,
              ]),
            ]),
          ]),
          'type': 'Polygon',
        }),
        'properties': dict({
          'Geometrietyp2DReferenz': '3000',
          'Grundrissaktualitaet': '2023-01-26',
          'bodenhoehe': '1100',
          'dachhoehe': '1000',
          'function': '31001_3070',
          'geschossanzahl': '1000',
          'id': 'DEHH_2cfcae78-56f5-4a77-b74d-51cce7cc3d3e_2_poly',
          'lage': '1000',
          'measuredHeight': '4.627',
          'roofType': '3100',
          'storeysAboveGround': '1',
          'surface_type': 'ground',
        }),
        'type': 'Feature',
      }),
      dict({
        'geometry': dict({
          'coordinates': list([
            list([
              list([
                10.316314,
                53.446814,
              ]),
              list([
                10.31598,
                53.446514,
              ]),
              list([
                10.315909,
                53.44654299999999,
              ]),
              list([
                10.316243,
                53.44684199999999,
              ]),
              list([
                10.316314,
                53.446814,
              ]),
            ]),
          ]),
          'type': 'Polygon',
        }),
        'properties': dict({
          'id': 'DEHH_d1c15e0c-707a-4734-abea-2a1f810c1890_2_poly',
          'surface_type': 'roof',
        }),
        'type': 'Feature',
      }),
      dict({
        'geometry': dict({
          'coordinates': list([
            list([
              list([
                10.316051,
                53.446486,
              ]),
              list([
                10.31598,
                53.446514,
              ]),
              list([
                10.316314,
                53.446814,
              ]),
              list([
                10.316386,
                53.44678499999999,
              ]),
              list([
                10.316051,
                53.446486,
              ]),
            ]),
          ]),
          'type': 'Polygon',
        }),
        'properties': dict({
          'id': 'DEHH_1fd3c3ef-fc46-4024-ac10-3b272044391d_2_poly',
          'surface_type': 'roof',
        }),
        'type': 'Feature',
      }),
    ]),
    'links': list([
    ]),
    'numberMatched': 3,
    'numberReturned': 3,
    'timeStamp': '2024-01-01T12:00:01Z',
    'type': 'FeatureCollection',
  })
# ---
# name: test_endpoint_with_limit
  dict({
    'features': list([
      'DEHH_2cfcae78-56f5-4a77-b74d-51cce7cc3d3e_2_poly',
    ]),
    'links': list([
      dict({
        'href': 'http://localhost/features?limit=1&startindex=1',
        'rel': 'next',
        'type': 'application/json',
      }),
    ]),
    'numberMatched': 5,
    'numberReturned': 1,
    'timeStamp': '2024-01-01T12:00:01Z',
    'type': 'FeatureCollection',
  })
# ---
# name: test_endpoint_with_limit_returns_prev_and_next
  dict({
    'features': list([
      'DEHH_5f54fbe2-c26d-4454-bf10-78b218e6d451_2_poly',
    ]),
    'links': list([
      dict({
        'href': 'http://localhost/features?limit=1&startindex=0',
        'rel': 'prev',
        'type': 'application/json',
      }),
      dict({
        'href': 'http://localhost/features?limit=1&startindex=2',
        'rel': 'next',
        'type': 'application/json',
      }),
    ]),
    'numberMatched': 5,
    'numberReturned': 1,
    'timeStamp': '2024-01-01T12:00:01Z',
    'type': 'FeatureCollection',
  })
# ---
